examples/test_util/EclFile.cpp

```diff
@@ -16,490 +16,807 @@
+
+                                                      if (p1 == -1) {
+                                                          p1 = val.find_first_of("-", 1);
+                                                          if (p1 > -1) {
+                                                              val = val.insert(p1,"E");
+                                                          } else {
+                                                              p1 = val.find_first_of("+", 1);
+
+                                                              if (p1 == -1) {
+                                                                  std::string message="In Routine Read readFormattedDoubArray, could not convert '" + val + "' to double.";
+                                                                  OPM_THROW(std::invalid_argument,message);
+                                                              }
+
+                                                              val = val.insert(p1,"E");
+                                                          }
+                                                      } else {
+                                                          val.replace(p1,1,"E");
+                                                      }
+
+                                                      return std::stod(val);
+                                                  };
+    return readFormattedArray<double>(fileH, size, f);
+}
+
+} // anonymous namespace
+
+
+EclFile::EclFile(const std::string& filename) : inputFilename(filename)
+{
+    std::fstream fileH;
+
+    formatted = isFormatted(filename);
+
+    if (formatted) {
+        fileH.open(filename, std::ios::in);
+    } else {
+        fileH.open(filename, std::ios::in |  std::ios::binary);
     }
-   
-    while (rest > 0) {   
- 
-        int num,dhead,dtail;
-
-        fileH.read((char *)&dhead, sizeof(dhead));
-        dhead=flipEndianInt(dhead);
-   
-        num=dhead/sizeOfElement;
-
-        if ((num > maxNumberOfElements) || (num < 0)){
-            std::string message="This ?? Error reading binary logi data, inconsistent header data or incorrect number of elements";
-            throw std::runtime_error(message);
+
+    if (!fileH) {
+        std::string message="Could not open file: " + filename;
+        OPM_THROW(std::runtime_error, message);
+    }
+
+    int n = 0;
+    while (!isEOF(&fileH)) {
+        std::string arrName(8,' ');
+        EIOD::eclArrType arrType;
+        int num;
+
+        if (formatted) {
+            readFormattedHeader(fileH,arrName,num,arrType);
+        } else {
+            readBinaryHeader(fileH,arrName,num,arrType);
         }
 
-        for (int i=0;i<num;i++){
+        array_size.push_back(num);
+        array_type.push_back(arrType);
 
-            bool value;
-            unsigned int intVal;
+        array_name.push_back(EIOD::trimr(arrName));
+        array_index[array_name[n]] = n;
```

Comment:
Array_index is a `std::map<std::string, int>`.  Does this mean that if a name repeats&mdash;as would be the case for the header arrays (e.g., `INTEHEAD`) of a unified restart file&mdash;that `array_index` records only the last occurrence of the array?

Reply:
Yes that is correct, and this is the one of the main reasons why we have to make specialized classes for unified restart files, summary files, and rft files. In the restart example below, your will get pressure from restart file SPE1CASE1 for report step number 12 (SEQNUM=12)

ERst rst1("SPE1CASE1.UNRST");
loadReportStepNumber(12);
rst1.getRst("PRESSURE", 12);

In program compareECL only INIT files are compared with the using EclFile.  Feel free to call me if you have other questions, or if you would like to discuss how this is implemented